### PR TITLE
adios2 2.5.0 (new formula)

### PR DESCRIPTION
**Pull request by ax3l**
~~MPI~~ HDF5 disabled since no parallel HDF5 is available in core.

cc-ing upstream maintainers @williamfgc @pnorbert @chuckatkins 

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by chuckatkins** [comment]
Would it instead be better to have MPI as a formula option and disable the hdf5 dependency if MPI is enabled?

**Comment by chuckatkins** [comment]
fwiw I don't advocate all the other options, but just two fixed configurations of with and without mpi.

File: Formula/adios2.rb
Comment (williamfgc) [inline Formula/adios2.rb]: @ax3l just curious about this dependency, adios2 doesn't explicitly depend on it.

File: Formula/adios2.rb
Comment (williamfgc) [inline Formula/adios2.rb]: @ax3l my understanding is that libfabric is only called if SST is ON

File: Formula/adios2.rb
Comment (williamfgc) [inline Formula/adios2.rb]: @ax3l my preference would be to keep MPI ON

**Comment by williamfgc** [comment]
@ax3l thanks for putting this together, I've been asked for a homebrew distribution. I'll try to address some clang related issues for the next releases. Thanks!

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: Is head needed? Or is upstream going to test regularly against homebrew? Or are the formula submitters testing and using it? If not, then it should be removed. Basically we want to include head if it's used regularly, especially to allow upstream to test, e.g., nightly.

Comment (chuckatkins) [reply Formula/adios2.rb]: `master` is used pretty often by developers working directly with adios and rarely by applications as a dependency, so it depends on the audience.  I'm in the process now of adding macOS builds to our pr-driven CI so I'll be sure to have the specific homebrew configurations tested as part of that.

**Review by zbeekman — CHANGES_REQUESTED** [review]
Looking good so far! A few comments:

## Tests

In general I like how thorough the tests are! I've added suggestions to check for the existence of the files created by the tests.

Would it make sense to try to simplify the C++ and python source in the tests at all? It's OK if the answer is "no" but I'm wondering if it is possible to strike a better balance between sufficiently testing ADIOS is capable of being built and linked against correctly while reducing complexity/maintenance burden of lots of code in the `test do` block.

## RE: HDF5 vs MPI

I'd have to feel this one out with the other maintainers, but maybe it would be worth adding a new version of HDF5 that has parallel IO enabled to Homebrew? For now, let's not count on this and work under the assumption that we have to choose one or another.

### HDF5 pros:

- can open data sets created with HDF5 locally for post-processing & visualization
- can write data sets (maybe part of a required pre-processing step) w/ HDF5 which may be required by certain applications

### MPI pros:

- Local development and testing of parallel applications
- Marginal speedups
  - I don't anticipate much production use of ADIOS installed via Homebrew
  - Some desperate users may install it via Homebrew for Linux, FKA linuxbrew
  - On macOS each process is most likely writing to the same storage device

### Conclusion

I'm inclined to favor HDF5 over MPI here for the reasons listed above if the anticipated use case on macOS is working with (post & pre -processing, visualization, etc.) ADIOS generated data.

If, however, the most likely use case is local development of parallel software relying on ADIOS for IO, then maybe MPI makes more sense.

## Dependencies

Please verify that `catch2` and `libfabric` are actually required dependencies, as noted by @williamfgc 

Also, please remove spurious comments in the dependencies section.

File: Formula/adios2.rb
Comment (zbeekman) [reply Formula/adios2.rb]: Awesome! For OpenCoarrays I have nightly CI tests against master built via `brew install --HEAD opencoarrays`. (They often fail though, due to issues with Travis-CI... so choose your CI provider carefully.) I am strongly in favor of keeping `head` for this case. If you setup nightly testing with `brew install --HEAD adios` and `brew test --HEAD adios` (and maybe throw a `brew audit` in there for good measure) then it will:

- Notify you if changes to upstream dependencies break ADIOS
- Let you know if some subtle change broke ADIOS working correctly with Homebrew
- Make it easy/seamless for Homebrew version bumps of ADIOS.

**Comment by zbeekman** [comment]
@chuckatkins 
> Would it instead be better to have MPI as a formula option and disable the hdf5 dependency if MPI is enabled?

We need to pick one for now. I may look into whether or not we could create a parallel HDF5 formula in Homebrew... but let's not wait on this. If that comes to fruition, then we can update the ADIOS formula at that time.

**Comment by chuckatkins** [comment]
> @chuckatkins
> 
> > Would it instead be better to have MPI as a formula option and disable the hdf5 dependency if MPI is enabled?
> 
> We need to pick one for now. I may look into whether or not we could create a parallel HDF5 formula in Homebrew... but let's not wait on this. If that comes to fruition, then we can update the ADIOS formula at that time.

Given that, an mpi enabled adios2 without hdf5 is a much more common use case than a serial adios2 with hdf5.

**Comment by chuckatkins** [comment]
> Would it make sense to try to simplify the C++ and python source in the tests at all? It's OK if the answer is "no" but I'm wondering if it is possible to strike a better balance between sufficiently testing ADIOS is capable of being built and linked against correctly while reducing complexity/maintenance burden of lots of code in the `test do` block.

@ax3l Rather that inline source, you could directly use the dummy code directly in the source tree used by the install tests: https://github.com/ornladios/ADIOS2/tree/master/testing/install

You could build and run them against the Brew installation.  It doesn't test io functionality but does test that basic configure-build-run works correctly.

File: Formula/adios2.rb
Comment (chuckatkins) [inline Formula/adios2.rb]: The flags used here might be different, especially with MPI in the mix.  The recommended practice is to use the output of the installed `adios2-config --cxx`, which can be passed directly to the compiler and is how an end user with makefiles would use it.

**Comment by chuckatkins** [comment]
Great work btw!  Thanks for getting this going!

**Comment by zbeekman** [comment]
> You could build and run them against the Brew installation. It doesn't test io functionality but does test that basic configure-build-run works correctly.

@ax3l: This is a good idea. You can "install" example/test code into the pkgshare directory and then test it in the `test do` block. See, e.g., `brew cat datetime-fortran` for a simple example of doing this.

**Comment by ax3l** [comment]
> Would it instead be better to have MPI as a formula option and disable the hdf5 dependency if MPI is enabled?

@chuckatkins Unfortunately, `homebrew-core` does not support options (variants) due to lack of CI: https://docs.brew.sh/Formula-Cookbook#specifying-other-formulae-as-dependencies

File: Formula/adios2.rb
Comment (ax3l) [reply Formula/adios2.rb]: ups, good catch. I copied from my openPMD-api recipe. Will remove!

File: Formula/adios2.rb
Comment (ax3l) [reply Formula/adios2.rb]: yes, SST is on here:
```
    Build Options:
      Blosc    : ON
      BZip2    : ON
      ZFP      : OFF
      SZ       : OFF
      MGARD    : OFF
      PNG      : ON
      MPI      : OFF
      DataMan  : ON
      Table    : OFF
      SSC      : ON
      SST      : ON
      DataSpaces: OFF
      ZeroMQ   : ON
      HDF5     : ON
      Python   : ON
      Fortran  : ON
      SysVShMem: ON
      Profiling: ON
      Endian_Reverse: OFF
    RDMA Transport for Staging: Availabl
```

File: Formula/adios2.rb
Comment (ax3l) [reply Formula/adios2.rb]: yes works for me as well and I thought about this, too.

File: Formula/adios2.rb
Comment (chuckatkins) [reply Formula/adios2.rb]: Note too that if you're using an mpi enabled adios2 then the flags from `adios2-config` assume you're using the mpi compiler wrappers so any flags added by `mpic++` are stripped from the output.

Comment (ax3l) [reply Formula/adios2.rb]: Okay, sure! So probably something like:
```suggestion
    adios2_config_flags = `adios2-config --cxx`.chomp.split
    system ENV.cxx, "write.cpp", *adios2_config_flags
```

File: Formula/adios2.rb
Comment (ax3l) [reply Formula/adios2.rb]: I think we can add a `hdf5-mpi` and `adios2-mpi` formula later on as well.
@zbeekman do you know how we could add such additional formulas without much duplication and with proper conflicts? Can be done as a follow-up.

**Comment by ax3l** [comment]
@zbeekman thank you a lot for the review and hints!

Re:MPI probably the easiest solution so far is to keep it serial, since I want to combine both HDF5 and ADIOS2 in a follow-up package. As written inline above: do you think it's easy to add - in a follow-up - `hdf5-mpi` and `adios2-mpi` formulas? I wonder if we can re-use code by deriving the packages and if we can express conflicts between the serial and mpi-parallel package?

@chuckatkins I am looking at https://github.com/ornladios/ADIOS2/tree/master/testing/install but the test cases are super small and do not use much of the API. Shall we add a little write example there as well? Or take something from https://github.com/ornladios/ADIOS2/tree/master/testing/adios2/bindings instead?

**Comment by zbeekman** [comment]
> Re:MPI probably the easiest solution so far is to keep it serial, since I want to combine both HDF5 and ADIOS2 in a follow-up package. As written inline above: do you think it's easy to add - in a follow-up - `hdf5-mpi` and `adios2-mpi` formulas?

In general, we need a very compelling reason to create additional packages. (See, for example real vs complex petsc). I think that there is zero chance that a second ADIOS package would be accepted. I agree with @chuckatkins that, in general,
> adios2 without hdf5 is a much more common use case than a serial adios2 with hdf5

with the caveat that we're talking mainly about macOS applications & users here. I'm inclined to defer to upstream here, and recommend you drop HDF5 in favor of MPI.

HDF5 is a popular enough package (close to 200k installs YTD, ~40k upon request YTD) that users---software developers---and parallel applications might benefit from having an MPI-IO enabled version. As such, if that package is introduced in the future, then we can modify the *single* ADIOS formula to turn on HDF5 and MPI.

@ax3l 
> I am looking at [ornladios/ADIOS2:testing/install@`master`](https://github.com/ornladios/ADIOS2/tree/master/testing/install) [](/ornladios/ADIOS2/tree/HEAD@{2019-10-30T21:17:21Z}/testing/install) but the test cases are super small and do not use much of the API. Shall we add a little write example there as well?

When writing tests for homebrew, the focus should be on testing that the installation is valid and reasonably functional. As such, for an IO library, if you can test out compiling and linking to it (especially to write or read a file) than that should be sufficient. We do not want or need large coverage of the API. If the package includes utility scripts to aid users in building & linking against the library, and these are the most common way that users do this, then you should test these too, especially since homebrew sanitizes the build and install environment so you may get unanticipated runtime results unless the compile/link utility scripts/tools are written in a very disciplined way.

**Comment by ax3l** [comment]
> > adios2 without hdf5 is a much more common use case than a serial adios2 with hdf5
>
> with the caveat that we're talking mainly about macOS applications & users here. I'm inclined to defer to upstream here, and recommend you drop HDF5 in favor of MPI.

Sounds reasonable to me as well. Yet, I want to link to both HDF5 and ADIOS2 in a downstream library, and using one with MPI and one without makes them mutually exclusive dependencies for my use case.

If there is a likely chance for an MPI-parallel package as you suggest than this sounds great to me. Will then already anticipate and write an MPI-parallel ADIOS formula herein.

> As such, for an IO library, if you can test out compiling and linking to it (especially to write or read a file) than that should be sufficient.

Hm, this sounds to me like I should keep my tests that actually do call some minimal write functionality and add the install sanity tests that do not read and write at all on top, as suggested by @chuckatkins . Or do you two advise otherwise? Definitely can also modify to solely keep one or another.

**Comment by ax3l** [comment]
Re:MPI have to submit a formula for the popular [mpi4py](https://mpi4py.readthedocs.io) package first in order to enable the MPI build.

**Comment by chuckatkins** [comment]
> > > adios2 without hdf5 is a much more common use case than a serial adios2 with hdf5
> > 
> > 
> > with the caveat that we're talking mainly about macOS applications & users here. I'm inclined to defer to upstream here, and recommend you drop HDF5 in favor of MPI.
> 
> Sounds reasonable to me as well. Yet, I want to link to both HDF5 and ADIOS2 in a downstream library, and using one with MPI and one without makes them mutually exclusive dependencies for my use case.
> 
> If there is a likely chance for an MPI-parallel HDF5 formula as you suggest than this sounds great to me. Will then already anticipate and write an MPI-parallel ADIOS formula herein.
> 
> > As such, for an IO library, if you can test out compiling and linking to it (especially to write or read a file) than that should be sufficient.
> 
> Hm, this sounds to me like I should keep my tests that actually do call some minimal write functionality and add the install sanity tests that do not read and write at all on top, as suggested by @chuckatkins . Or do you two advise otherwise? Definitely can also modify to solely keep one or the other.

I still think it's problematic to keep large inline test code in the formula, but I get that you want just a *little* more functionality in the test than compile / link.  To that end, I suggest using one of the basic "hello world" examples in `examples/hello/bpWriter` from the source tree.  You get the benefits of:
- a simpler formula
- using the example code that's maintained with the repo so it should be guaranteed to work
- more that a basic compile test as the example also does a basic write with the api
- a pre-baked example for all supported language bindings: C, C++, Fortran, and Python

**Comment by williamfgc** [comment]
@zbeekman thanks for tracking this and providing your feedback, it's very helpful.

> Are we sure mpi4py is a hard dependency for adios with MPI enabled?

Not a hard depency for adios2 per-se, but for the adios2 Python bindings (they build on top of adios2).
The dependency relationship is as follows:
- `adios2+mpi+python` requires `mpi4py`, `numpy`, `python`
- `adios2+python` requires `numpy`, `python`

> Homebrew doesn't accept formulae that are python libraries (at least not without some useful command line tool or program).

What's the recommended way to install Python bindings from an existing formulae? Thanks!

**Comment by chuckatkins** [comment]
> Do either of you know if `mpi4py` has a shot at working correctly when you simply `pip install mpi4py`? In cases where this is _NOT_ true, i.e., pip can't produce a useful library installation, we will accept formulae for stand alone python libraries/modules.

So long as the MPI compiler wrappers are available (mpicc, etc.) In the path then `pip install mpi4py` should be just fine.

**Comment by ax3l** [comment]
Thanks for your feedback you all!

I applied the changes and `mpi4py` only depends on `MPI` and `Python` as vendored resource.
The only step that is missing is to tell ADIOS2' CMake the location of the freshly installed `mpi4py` during install so it can find it in `brew/Cellar/adios2/2.5.0/libexec/lib/python3.7/site-packages/` during the `install` block.

I wonder though: since homebrew vendors the Cython package `numpy` if the Cython package `mpi4py` would not be a candidate for a regular Formula as well.

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: If you think building mpi4py with cython is a good idea then would adding this make sense? (And the extra logic required to use cython in the install block.)

```suggestion
  depends_on "cmake" => :build
  depends_on "cython" => :build
```

Or would it require a stand alone mpi4py formula?

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: This looks like it should work. I'm not sure if you need to pass anything to CMake, it should query the python path, no?

**Comment by zbeekman** [comment]
> I wonder though: since homebrew vendors the Cython package `numpy` isn't the Cython package `mpi4py` another good candidate for a regular Formula as well?

Maybe. I'm personally amenable to the idea. I'm not sure whether or not others will agree. A quick search looks like no other formulae are vendoring mpi4py though, so it may not be that popular.

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: These tests look great! I'm not sure we need to test mpi4py import functionality but I'm happy enough to leave it as it is.

Comment (ax3l) [reply Formula/adios2.rb]: yes, have this intentionally right now because the `import mpi4py` line does not yet work.

**Comment by ax3l** [comment]
> Maybe. I'm personally amenable to the idea. I'm not sure whether or not others will agree. A quick search looks like no other formulae are vendoring mpi4py though, so it may not be that popular.

Sounds great. This might be a chicken-egg problem here, `mpi4py` is the only maintained python binding for MPI currently and all popular packages such as `h5py` use it. Will submit an independent PR for it.

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: I don't think you need this now that `mpi4py` is in homebrew

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]:
```suggestion
  depends_on "mpi4py"
  depends_on "numpy"
```

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: Can get rid of the resource now 😄

**Comment by ax3l** [comment]
Ok, high_sierra builds fail on a bug I already fixed in `master`: https://github.com/ornladios/ADIOS2/pull/1805/files

Can I specify this as patch to the Formula?

**Comment by ax3l** [comment]
... and it was, dear robot ;)

File: Formula/adios2.rb
Comment (zbeekman) [inline Formula/adios2.rb]: @ax3l you rock! This is exactly what we need here.

The only thing that it's missing is a comment indicating when to remove the path, and a very very brief description of the problem. I presume it should be removed in the next major release. Could you please add a comment providing these two details?

**Review by zbeekman — CHANGES_REQUESTED** [review]
Almost there. I think all that's missing now is a comment for the patch.

**Comment by zbeekman** [comment]
Oh, also to squash the commits, if you can. Please have a single commit. You can do this when you add the comment for the patch.

**Comment by ax3l** [comment]
No problem, cleaned, commented and squashed.
Thank you everyone for the input, added you as co-authors of the commit for credit :sparkles:

**Review by zbeekman — APPROVED** [review]
LGTM!

**Comment by zbeekman** [comment]
Thanks for your contribution to Homebrew! 🎉 🥇 

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you @ax3l!!!

**Pull request closed without merging**
